prototype/mobile_endpoint/models.py

```diff
@@ -0,0 +1,191 @@
+            new = True
+
+        self.domain = generic.domain
+        self.received_on = generic.received_on
+        self.user_id = generic.metadata.userID
+        self.form_json = generic.to_json()
+        self.type = generic.doc_type
+        return new, self
+
+    def __repr__(self):
+        return (
+            "FormData("
+                "id='{f.id}', "
+                "domain='{f.domain}', "
+                "received_on='{f.received_on}', "
+                "user_id='{f.user_id}', "
+                "type='{f.type}')"
+        ).format(f=self)
+
+
+class CaseData(db.Model, ToFromGeneric):
+    __tablename__ = 'case_data'
+    id = db.Column(UUID(), primary_key=True)
+    domain = db.Column(db.Text(), nullable=False)
+    closed = db.Column(db.Boolean(), default=False, nullable=False)
+    owner_id = db.Column(UUID(), nullable=False)
+    server_modified_on = db.Column(db.DateTime(), nullable=False)
+    case_json = db.Column(JSONB(), nullable=False)
+
+    forms = db.relationship("FormData", secondary=case_form_link, backref="cases")
+
+    def to_generic(self):
+        if self.case_json:
+            generic = CommCareCase.wrap(self.case_json)
+        else:
+            generic = CommCareCase()
+
+        generic.id = self.id
+        generic.owner_id = self.owner_id
+        generic.closed = self.closed
+        generic.domain = self.domain
+        generic.server_modified_on = self.server_modified_on
+        for index in self.indices:
+            generic.indices.append(index.to_generic())
+        generic._self = self
+        return generic
+
+    @classmethod
+    def from_generic(cls, generic, **kwargs):
+        if hasattr(generic, '_self'):
+            self = generic._self
+            new = False
+        else:
+            self = cls(id=generic.id)
+            new = True
+
+        json = generic.to_json()
+        self.domain = json.pop('domain')
+        self.server_modified_on = json.pop('server_modified_on')
+        self.owner_id = json.pop('owner_id')
+        self.closed = json.pop('closed')
+        json.pop('indices')  # drop indices since they are stored separately
+        self.case_json = json
+        return new, self
+
+    def __repr__(self):
+        return (
+            "CaseData("
+                "id='{c.id}', "
+                "domain='{c.domain}', "
+                "closed={c.closed}, "
+                "owner_id='{c.owner_id}', "
+                "server_modified_on='{c.server_modified_on}')"
+        ).format(c=self)
+
+db.Index('ix_case_data_domain_owner', CaseData.domain, CaseData.owner_id)
+db.Index('ix_case_data_domain_closed_modified', CaseData.domain, CaseData.closed, CaseData.server_modified_on)
+
+
+class CaseIndex(db.Model, ToFromGeneric):
+    __tablename__ = 'case_index'
+    case_id = db.Column(UUID(), db.ForeignKey('case_data.id'), primary_key=True)
+    identifier = db.Column(db.Text(), primary_key=True)
+    referenced_id = db.Column(UUID(), db.ForeignKey('case_data.id'))
```

Comment:
do we want to include domain here as well? I know it's redundant, but it's a pretty basic silo and nice to index by, and for validation.